### PR TITLE
Invalidate PR build results after 3 hours

### PR DESCRIPTION
**Pull request by SaemJeon**
_The following is our draft announcement for this change:_

We have just made a significant change to our GitHub PR build automation in an effort to reduce build machine load and CI queue wait times.  **TLDR: We're going back to how things worked in Azure DevOps - you may occasionally have to click _Update Branch_ to requeue builds if your PR gets too stale.**

Essentially, PRs will no longer automatically be updated, and after 3 hours successful builds checks will be invalidated (if the PR branch is out of date).  For more details, please see below.

This guide will explain the updated pull request workflow which we expect should save a lot of unnecessary build machine time and thus reduce build queue wait times.

## Previous workflow
The imodeljs GitHub repo uses a tool called [mergify](https://mergify.io/) to automate a number of common merge tasks.  Until now, we have used mergify's [strict option](https://docs.mergify.io/actions/merge/#strict-merge) to require that every PR branch is 100% up-to-date with its target branch before it is allowed to merge.  In addition, mergify has automatically updated PR branches when there was an update to the target branch (master, release/\*, etc).  This has worked to ensure nothing stale ever merges and breaks the master branch, but it is causing way too many unnecessary builds.  Basically, without any smart queuing, we have PRs merging one-at-a-time, each one causing _all_ other PR branches to update and rerun builds.

## New workflow
We have disabled mergify's strict mode and auto-updating.  Instead, every open (non-draft) PR is now subject to a **3 hour policy**.  Basically, 3 hours after successfully passing all build checks, PRs that are out-of-date with their target branch will have their build checks invalidated.  This means the following PRs will **not** be affected:
- Draft PRs
- Any PR that is successfully reviewed, built, and merged in < 3 hours
- Any PR still up-to-date with the target branch
    - Even if the PR is older than 3 hours, the new workflow checks to see if the PR is up-to-date with the target branch by comparing the tip of the target brach SHA with the base SHA.  Only PRs > 3 hours old _and_ not up-to-date with the target branch are invalidated.
    - In other words, a PR's builds will not be unnecessarily invalidated if no other PRs are merged ahead of them.
    
Since only out-of-date PR builds are ever invalidated, **you can always requeue builds using the update branch button**:
image

### CODE REVIEW

File: .github/workflows/invalidatePRs.sh [hunk 1]
```diff
@@ -39,19 +35,58 @@ function log() {
...
+  if [[ $ShaToCompare == "" ]]
+  then
+    refCommit=$(curl -s \
+              -H "Accept: application/vnd.github.v3+json" \
+              -H "Authorization: token $oauth" \
+              https://api.github.com/repos/imodeljs/imodeljs/commits/${ref})
+    refHeadCommit=$(echo ${refCommit} | jq -r '. | @base64')
+    refHeadSha=$(_jq  ${refHeadCommit} '.sha')
+    listSHA[${ref}]=${refHeadSha}
+    ShaToCompare=$refHeadSha
```
Comment (wgoehrig) [inline .github/workflows/invalidatePRs.sh:75]: nit: this would be more readable if the variable name casing was consistent (i.e., rename this to `shaToCompare`)

**Comment by wgoehrig** [comment]
@pmconne @kabentley @ColinKerr @grigasp 

Can you please review the announced PR workflow changes in this PR's description?  We're currently planning to roll this out Thursday (4/15), and I just wanted to get some feedback from some of you all before we pull the trigger.

**Comment by pmconne** [comment]
@wgoehrig I already told @calebmshafer I am on board.

**Comment by grigasp** [comment]
> @pmconne @kabentley @ColinKerr @grigasp
> 
> Can you please review the announced PR workflow changes in this PR's description? We're currently planning to roll this out Thursday (4/15), and I just wanted to get some feedback from some of you all before we pull the trigger.

One piece that's not completely clear to me is _"3 hours after successfully passing all build checks"_ - if one of the PR builds takes several hours to complete (including all the restarts due to sporadic build failures), I assume we start counting 3 hours only after that last build completes - that means the PR could stay be valid for much longer than 3 hours, right?

File: .github/workflows/invalidatePRs.sh [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      log "  Skipping $(_jq  ${status} '.context') it was updated within the past 3 hours."
+      continue
+    fi
+
+    echo "  The PR $(_jq ${pr} '.title') has build $(_jq  ${status} '.context') that is older than 3 hours. Invalidating..."
+
+    updateUrl=https://api.github.com/repos/imodeljs/imodeljs/statuses/$(_jq ${pr} '.head.sha')
+    target_url=$(_jq  ${status} '.target_url')
+
+    # curl \
```
Comment (wgoehrig) [inline .github/workflows/invalidatePRs.sh:122]: Need to uncomment this before we merge.

File: .github/workflows/invalidatePRs.sh [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    log "  Skipping due to state=$(_jq ${pr} '.state') and draft=$(_jq ${pr} '.draft')."
+    continue
+  fi
+
+  ref=$(_jq ${pr} '.base.ref')
+  sha=$(_jq ${pr} '.base.sha')
+
+  shaToCompare=${listSha[${ref}]}
+
+  # Keep the list of branch head sha
```
Comment (calebmshafer) [inline .github/workflows/invalidatePRs.sh:65]: @saemjeon can you update this comment? It's a bit misleading about what is happening here

File: .github/workflows/invalidatePRs.sh [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+              -H "Authorization: token $oauth" \
+              https://api.github.com/repos/imodeljs/imodeljs/commits/${ref})
+    refHeadCommit=$(echo ${refCommit} | jq -r '. | @base64')
+    refHeadSha=$(_jq  ${refHeadCommit} '.sha')
+    listSha[${ref}]=${refHeadSha}
+    shaToCompare=$refHeadSha
+  fi
+
+  # Compare base sha with the target branch head sha
+  log "  Head sha at ${ref} branch is ${shaToCompare} and base sha is ${sha}"
```
Comment (calebmshafer) [inline .github/workflows/invalidatePRs.sh:79]: Instead of `head sha` can you make this `Target branch, ${ref}, commit sha is ${shaToCompare} and PR is on sha ${sha}`?  Or something along those lines just to make it more clear that it is the target branch that is further ahead.

File: .github/workflows/invalidatePRs.sh [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  lastPushedTime=$(_jq ${branchLatestCommit} '.commit.commit.author.date')
+
+  log "  PR created at ${prCreationTime}."
+  log "  Last pushed at ${lastPushedTime}."
+
+  currentTime=$(date -u +"%Y-%m-%dT%H:%M:%SZ")
+  createdTooOld=$(node -e "const lutPlus3=new Date('$prCreationTime'); lutPlus3.setHours(lutPlus3.getHours()+3); console.log((new Date('$currentTime') - lutPlus3) > 0)")
+  pushedTooOld=$(node -e "const lutPlus3=new Date('$lastPushedTime'); lutPlus3.setHours(lutPlus3.getHours()+3); console.log((new Date('$currentTime') - lutPlus3) > 0)")
+  if [[ $createdTooOld == false || $pushedTooOld == false ]]; then
+    log "  Skipping since this PR has been created or last pushed within the past 3 hours."
```
Comment (calebmshafer) [inline .github/workflows/invalidatePRs.sh:103]: `last pushed` -> `updated`

**Comment by calebmshafer** [comment]
> > @pmconne @kabentley @ColinKerr @grigasp
> > Can you please review the announced PR workflow changes in this PR's description? We're currently planning to roll this out Thursday (4/15), and I just wanted to get some feedback from some of you all before we pull the trigger.
> 
> One piece that's not completely clear to me is _"3 hours after successfully passing all build checks"_ - if one of the PR builds takes several hours to complete (including all the restarts due to sporadic build failures), I assume we start counting 3 hours only after that last build completes - that means the PR could stay be valid for much longer than 3 hours, right?

@grigasp the way it is currently implemented, each build is validated independently.  If the PR has all but one build pass and it takes over 3 hours to get that last one to completed due to flakiness, those other builds will be invalidated.  Making the entire PR in an invalid state, that is all predicated on if the target branch is updated in that time period.

File: .github/workflows/invalidatePRs.sh [hunk 6]
```diff
@@ -0,0 +1,141 @@
+#!/bin/bash
+
+# GitHub workflow handles invalidating Pull Requests after 3 hours.
+#
+# It will only invalidate open, non-draft pull requests and mark each status
```
Comment (calebmshafer) [inline .github/workflows/invalidatePRs.sh:5]: Can you add a bit more description here about the possible cases. Similar to my response to Grigas, we should write out how it works in those explicit cases.

Comment (saemjeon) [reply .github/workflows/invalidatePRs.sh]: Added some description and edited some comments as per your comment. Let me know what you think of the description (possible cases) I added. I am not sure if I accounted all the explicit cases or not.

**Comment by calebmshafer** [comment]
@saemjeon reviewing a test run of the action and it seems to be only invalidating the first build it finds for a given PR.  Not quite sure what's going on, if it's only doing 1 or skipping Docs for a reason.  If it's just doing one, I think that is probably fine since it'd require an "Update branch" which would re-queue them all. If it's the Docs for some reason, that one should be fixed.

I was checking this [build](https://github.com/imodeljs/imodeljs/runs/2530556997?check_suite_focus=true) and looking at this [PR](https://github.com/imodeljs/imodeljs/pull/1350). It has two builds that should be invalidated [Integration](https://dev.azure.com/bentleycs/iModelTechnologies/_build/results?buildId=877831&view=results) and [Docs](https://dev.azure.com/bentleycs/iModelTechnologies/_build/results?buildId=877829&view=results) but only the Integration one was.

File: .github/workflows/invalidate-prs.yaml [hunk 7]
```diff
@@ -0,0 +1,26 @@
+# This workflow invalidates the builds of pull requests that more than 3 hours old.
+#
+
+name: "Invalidate PRs after 3 hours"
+
+on:
+  workflow_dispatch:
+  schedule:
+    # * is a special character in YAML so you have to quote this string
```
Comment (calebmshafer) [inline .github/workflows/invalidate-prs.yaml:9]: Can you add to this comment that the pattern defines that it runs ever 15 minutes.

**Review by calebmshafer — COMMENTED** [review]
> For more details please see this wiki article.

Where were you thinking we post this wiki article?

Running some additional tests now, and then we can get this merged after uncommenting the section mentioned by @wgoehrig

File: .github/workflows/invalidate-prs.yaml [hunk 8]
```diff
@@ -0,0 +1,26 @@
+# This workflow invalidates the builds of pull requests that more than 3 hours old.
+#
+
+name: "Invalidate PRs after 3 hours"
+
+on:
+  workflow_dispatch:
```
Comment (calebmshafer) [inline .github/workflows/invalidate-prs.yaml:7]: Can you add an input to the dispatch for testing purposes?

It'd look something like this

```yaml
inputs:
      debug:
        description: 'Enable debug logging'     
        required: false
```

with the script task using it via `${{ github.event.inputs.debug }}` as an environment variable.

I'm looking at [this](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputs) page for reference.

File: .github/workflows/invalidate-prs.yaml (hunk 7 above)
Comment (SaemJeon) [reply .github/workflows/invalidate-prs.yaml:9]: I think @wgoehrig mentioned about internal wiki page.

File: .github/workflows/invalidatePRs.sh (hunk 2 above)
Comment (calebmshafer) [reply .github/workflows/invalidatePRs.sh:145]: @SaemJeon you should be good to uncomment this. I'll make sure the token is setup correctly.

**Comment by calebmshafer** [comment]
@SaemJeon lets merge this PR in today, after that last section is uncommented. After that, on Monday, we'll need a separate PR to turn of mergify and I'll remove the strictness of always having to keep the branch up-to-date.

**Review by calebmshafer — APPROVED** [review]
